README.md
[![Build Status](https://travis-ci.org/goodybag/node-pg-transaction.png)](https://travis-ci.org/goodybag/node-pg-transaction)

Make it easier to write transactions for PostgreSQL using [node-postgres](https://github.com/brianc/node-postgres).

The callback and event emitter styles both work.  
**Note:** if you use a callback, then the error event won't be emitted. This is consistent with node-postgres.

##Installation:

To install the most recent release from npm, run:

```
npm install pg-transaction
```

## Methods:

```javascript
begin(callback);
query(config, values, callback); // this is pg.Client.query from node-postgres
savepoint(savepoint, callback);
release(savepoint, callback);
rollback(savepoint, callback);
commit(callback);
```

## Events:

- row
- end
- error

## Example:

```javascript

/**
 * Module dependencies
 */

var
  // PostgreSQL modules
  pg = require('pg')
, Transaction = require('pg-transaction')

  // Configuration stuff
, connectionString = process.env['PG_CON'] || ''
;

var die = function(err){
  if (err) throw err;
};

var client = new pg.Client(connectionString);
client.connect();

client.query("CREATE TEMP TABLE beatles(name varchar(10), height integer, birthday timestamptz)");

var tx = new Transaction(client);
tx.on('error', die);

tx.begin();
tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
tx.savepoint('savepoint1');
tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
tx.rollback('savepoint1'); // all statements after savepoint1 are undone (John will not be inserted)
tx.release('savepoint1'); // can no longer use savepoint1 as a point to rollback to
tx.commit();

client.query("SELECT COUNT(*) AS count FROM beatles", function(err, result){
  if (err) return die(err);
  console.log(result.rows[0].count); // 1
  client.end();
});
```

## Contributors:

- [Lalit Kapoor](https://github.com/lalitkapoor)

## Special thanks to the following:

- [Brian M. Carlson](https://github.com/brianc)


CHANGELOG.md
## Change Log

### v1.0.2 (2014/01/19 01:40 +00:00)
- [#2](https://github.com/goodybag/node-pg-transaction/pull/2) add test to confirm user defined type parsers work correctly (@lalitkapoor)

- [#4](https://github.com/goodybag/node-pg-transaction/pull/4) create additional docs using dox (@lalitkapoor)

- [#5](https://github.com/goodybag/node-pg-transaction/pull/5) major version bump (@brianc)

- [#6](https://github.com/goodybag/node-pg-transaction/pull/6) support postgres 2.x (@lalitkapoor)

test/index.js

/**
 * Module dependencies
 */

var
  pg = require('pg')
, async = require('async')
, should = require('should')
, Transaction = require('../')
, assert = require('assert')
;

var
  connectionStr = process.env['PG_CON'] || ''
;

describe('transaction', function(){
  beforeEach(function(){
    this.client = new pg.Client(connectionStr);
    this.client.connect();
    this.client.query("CREATE TEMP TABLE beatles(name varchar(10), height integer, birthday timestamptz, large bigint)");
  });

  afterEach(function(done){
    this.client.on('end', done);
    this.client.end();
  });

  it('#commit - should exist in the database', function(done){
    var tx = new Transaction(this.client);
    tx.begin();
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
    tx.commit();
    this.client.query("SELECT * FROM beatles", function(err, result){
      if (err) throw err;
      result.rows.should.have.length(2);
      done();
    });
  });

  it('#rollback - should not exist in the database', function(done){
    var tx = new Transaction(this.client);
    tx.begin();
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
    tx.rollback();
    this.client.query("SELECT * FROM beatles WHERE name = $1", ['John'], function(err, result){
      if (err) throw err;
      result.rows.should.have.length(0);
      done();
    });
  });

  it('#savepoint - create a savepoint and rollback to it', function(done){
    var tx = new Transaction(this.client);
    tx.begin();
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
    tx.savepoint('test');
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
    tx.rollback('test');
    tx.commit();
    this.client.query("SELECT * FROM beatles", function(err, result){
      if (err) throw err;
      result.rows.should.have.length(1);
      done();
    });
  });

  it('#release - destroy a savepoint (cannot rollback to it anymore)', function(done){
    var tx = new Transaction(this.client);
    tx.on('error', function(err){
      throw err;
    });

    tx.begin();
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
    tx.savepoint('test');
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
    tx.release('test');
    tx.rollback('test', function(err){
      should.exist(err);
      err.code.should.equal('3B001');
    });
    tx.commit(done);
  });

  it('error in transaction', function(done){
    var tx = new Transaction(this.client);

    var successful = true;
    tx.on('error', function(err){
      successful = false;
    });

    tx.begin();
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
    tx.query("INSERT INTO dummy(name, height, birthday) values($1, $2, $3)", ['Bob', 68, new Date(1944, 10, 13)]);
    tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
    tx.commit();
    this.client.query("SELECT * FROM beatles", function(err, result){
      if (err) throw err;
      successful.should.equal(false);
      result.rows.should.have.length(0);
      done();
    });
  });

  it('recover from error in transaction by using a savepoint', function(done){
    var self = this;
    var tx = new Transaction(this.client);

    var checkDone = function(){
      self.client.query("SELECT * FROM beatles", function(err, result){
        if (err) throw err;
        result.rows.should.have.length(2);
        done();
      });
    }

    tx.begin(function(err){
      if (err) throw err;
      tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)], function(err){
        if (err) throw err;
        tx.savepoint('savepoint1', function(err){
          if (err) throw err;
          tx.query("INSERT INTO dummy(name, height, birthday) values($1, $2, $3)", ['Bob', 68, new Date(1944, 10, 13)], function(err){
            if (err) {
              tx.rollback('savepoint1', function(err){
                if (err) throw err;
                tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3) RETURNING *", ['John', 68, new Date(1944, 10, 13)], function(err, result){
                  assert.ifError(err);
                  assert(result, "should have a result");
                  assert(result.rows, "result should have rows");
                  result.rows.length.should.eql(1);
                  var john = result.rows[0];
                  john.name.should.eql('John');
                  john.height.should.eql(68);
                  john.birthday.getFullYear().should.eql(1944);
                  tx.commit(function(err){
                    if (err) throw err;
                    checkDone();
                  });
                });
              });
            } else {
              tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)], function(err){
                tx.commit(function(err){
                  if (err) throw err;
                  checkDone();
                });
              });
            }
          });
        });
      });
    });
  });

  it('#abort a failing transaction', function(done){
    var stage = null;
    var tx = new Transaction(this.client);
    async.series(
      {
        begin: function(callback){
          stage = 'begin';
          tx.begin(callback);
        }
      , i1: function(callback){
          stage = 'i1';
          tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)], callback);
        }
      , i2: function(callback){
          stage = 'i2';
          tx.query("INSERT INTO dummy(name, height, birthday) values($1, $2, $3)", ['Bob', 68, new Date(1944, 10, 13)], callback);
        }
      , i3: function(callback){
          stage = 'i3';
          tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)], callback);
        }
      , commit: function(callback){
          stage = 'commit';
          tx.commit(function(err){
            if (err) return callback(err);
            done();
          });
        }
      }
    , function(err, results){
        if('i2' === stage) should.exist(err);
        if (err) return tx.abort(function(err){
          if (err) throw err;
          done();
        });
      }
    );
  });

  it('should use custom defined type parsers', function(done){
    var originalTypeParser20 = pg.types.getTypeParser(20);
    pg.types.setTypeParser(20, String);
    var self = this;
    var tx = new Transaction(this.client);
    async.series(
      {
        begin: function(callback) { 
          tx.begin(callback);
        }
      , i1: function(callback) {
          callback();
          tx.query("INSERT INTO beatles(name, height, birthday, large) values($1, $2, $3, $4)", ['Ringo', 67, new Date(1945, 11, 2), '12345678901'], callback);
        }
      , commit: function(callback) { 
          tx.commit(callback);
        }
      , confirm: function(callback) {
          self.client.query("SELECT large FROM beatles WHERE name='Ringo'", function(err, result){
            if (err) return callback(err);
            result.rows.should.have.length(1);
            result.rows[0].large.should.be.a('string');

            // restore original parser
            pg.types.setTypeParser(20, originalTypeParser20);
            done();
          });
        }
      }
    , function(err, results) {
        if (err) {
          pg.types.setTypeParser(20, originalTypeParser20);
          throw err;
        }
      }
    );
  });
});


lib/index.js
/**
 * Module dependencies
 */

var
  pgClient = require('pg').Client
, EventEmitter = require('events').EventEmitter
, util = require('util')
;

var Transaction = module.exports = function(client,done) {
  this.client = client;
  this.__done = done;
};

util.inherits(Transaction, EventEmitter);

/**
 * Execute a query, re-emit the events that the client is receiving from this
 * EventEmitter
 */
Transaction.prototype.query = function(){
  var self = this;

  var query = this.client.query.apply(this.client, arguments);

  var callback = query.callback;

  if(callback == null){
    query.on('error', function(err){
      self.emit('error', err);
    });
  }
  return query;
};

/**
 * Start a transaction block
 * @param  {String}   transaction mode [optional] mode of transaction
 * @param  {Function} callback
 */
Transaction.prototype.begin = function(mode,callback){
  if(typeof(mode) === 'function'){
    callback=mode;
    mode=null;
  }
  var begin="BEGIN";
  if (mode) { begin+=" "+mode; }
  this.query(begin, callback);
};

/**
 * Define a new savepoint within the current transaction
 * @param  {String}   savepoint name of the savepoint
 * @param  {Function} callback
 */
Transaction.prototype.savepoint = function(savepoint, callback) {
  this.query('SAVEPOINT ' + savepoint, callback);
};

/**
 * Destroy a previously defined savepoint
 * @param  {String}   savepoint name of the savepoint
 * @param  {Function} callback
 */
Transaction.prototype.release = function(savepoint, callback) {
  this.query('RELEASE SAVEPOINT ' + savepoint, callback);
};

/**
 * Commit the current transaction
 * @param  {Function} callback
 */
Transaction.prototype.commit = function(callback){
  var self = this;

  this.query('COMMIT', function(err){
    if (callback) return callback(err);
    if (err) return self.emit('error', err);
  });
};

/**
 * Abort the current transaction or rollback to a previous savepoint
 * @param  {String}   savepoint [optional] name of the savepoint to rollback to
 * @param  {Function} callback
 */
Transaction.prototype.rollback = function(savepoint, callback){
  var self = this;

  if(typeof(savepoint) === 'function'){
    savepoint = null;
    callback = savepoint;
  }

  var query = (savepoint != null) ? 'ROLLBACK TO SAVEPOINT ' + savepoint : 'ROLLBACK';

  this.query(query, function(err){
    if (callback) return callback(err);
    if (err) return self.emit('error', err);
  });
};

/**
 * Abort the current transaction
 * @param  {Function} callback
 */
Transaction.prototype.abort = function(callback) {
  var self = this;

  this.query('ABORT TRANSACTION', function(err){
    if (callback) return callback(err);
    if (err) return self.emit('error', err);
  });
};

/**
 * Release current pg connection to pool
 */
Transaction.prototype.done = function() {
  if (this.__done) this.__done();
  this.__done=null;
};


examples/example.js
var
  pg = require('pg')
, Transaction = require('../')

, connectionString = process.env['PG_CON'] || ''
;

var die = function(err){
  if (err) throw err;
};

var client = new pg.Client(connectionString);
client.connect();

client.query("CREATE TEMP TABLE beatles(name varchar(10), height integer, birthday timestamptz)");

var tx = new Transaction(client);
tx.on('error', die);

tx.begin();
tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['Ringo', 67, new Date(1945, 11, 2)]);
tx.savepoint('savepoint1');
tx.query("INSERT INTO beatles(name, height, birthday) values($1, $2, $3)", ['John', 68, new Date(1944, 10, 13)]);
tx.rollback('savepoint1'); // all statements after savepoint1 are undone (John will not be inserted)
tx.release('savepoint1'); // can no longer use savepoint1 as a point to rollback to
tx.commit();

client.query("SELECT COUNT(*) AS count FROM beatles", function(err, result){
  if (err) return die(err);
  console.log(result.rows[0].count); // 1
  client.end();
});
